classes/plugins/PubIdPlugin.inc.php

```diff
@@ -32,333 +31,228 @@ function PubIdPlugin() {
-				if ($pubId == $existingPubId) return false;
-			}
+				break;
 
-			unset($objectsToCheck);
-		}
+			default:
+				$pubIdSuffix = String::strtolower($context->getAcronym($context->getPrimaryLocale()));
 
-		// We did not find any ID collision, so go ahead.
-		return true;
-	}
+				if ($issue) {
+					$pubIdSuffix .= '.v' . $issue->getVolume() . 'i' . $issue->getNumber();
+				} else {
+					$pubIdSuffix .= '.v%vi%i';
+				}
 
-	/**
-	 * Add the suffix element and the public identifier
-	 * to the object (issue, article, galley, supplementary file).
-	 * @param $hookName string
-	 * @param $params array ()
-	 */
-	function getAdditionalFieldNames($hookName, $params) {
-		$fields =& $params[1];
-		$formFieldNames = $this->getFormFieldNames();
-		foreach ($formFieldNames as $formFieldName) {
-			$fields[] = $formFieldName;
+				if ($article) {
+					$pubIdSuffix .= '.' . $article->getId();
+				}
+
+				if ($galley) {
+					$pubIdSuffix .= '.g' . $galley->getId();
+				}
 		}
-		$daoFieldNames = $this->getDAOFieldNames();
-		foreach ($daoFieldNames as $daoFieldName) {
-			$fields[] = $daoFieldName;
+		if (empty($pubIdSuffix)) return null;
+
+		// Costruct the pub id from prefix and suffix.
+		$pubId = $this->constructPubId($pubIdPrefix, $pubIdSuffix, $contextId);
+
+		if ($pubId && !$preview) {
+			$this->setStoredPubId($pubObject, $pubObjectType, $pubId);
 		}
-		return false;
+
+		return $pubId;
 	}
 
+	//
+	// Public API
+	//
 	/**
-	 * Return the object type.
-	 * @param $pubObject object
-	 *  (Issue, Article, PublishedArticle, ArticleGalley)
-	 * @return array
+	 * Exclude all issue objects (articles, galleys) from
+	 * assigning them the pubId or
+	 * clear pubIds of all issue objects (articles, galleys)
+	 * @param $exclude bool
+	 * @param $clear bool
+	 * @param $issue Issue
 	 */
-	function getPubObjectType($pubObject) {
-		$allowedTypes = array(
-			'Issue' => 'Issue',
-			'Article' => 'Article',
-			'ArticleGalley' => 'Galley',
-		);
-		$pubObjectType = null;
-		foreach ($allowedTypes as $allowedType => $pubObjectTypeCandidate) {
-			if (is_a($pubObject, $allowedType)) {
-				$pubObjectType = $pubObjectTypeCandidate;
-				break;
+	function excludeAndClearIssueObjects($exclude, $clear, $issue) {
+		$issueId = $issue->getId();
+		$articlePubIdEnabled = $this->isObjectTypeEnabled('Article', $issue->getJournalId());
+		$galleyPubIdEnabled = $this->isObjectTypeEnabled('ArticleGalley', $issue->getJournalId());
+		if (!$articlePubIdEnabled && !$galleyPubIdEnabled) return false;
+		$settingName = $this->getExcludeFormFieldName();
+		$pubIdType = $this->getPubIdType();
+		$articleDao =& DAORegistry::getDAO('ArticleDAO');
+		$publishedArticleDao =& DAORegistry::getDAO('PublishedArticleDAO');
+		$publishedArticles = $publishedArticleDao->getPublishedArticles($issueId);
+		foreach ($publishedArticles as $publishedArticle) {
+			if ($articlePubIdEnabled) {
+				if ($exclude && (!$publishedArticle->getStoredPubId($pubIdType) || $clear)) {
```

Comment:
Guessing a lot of this exclusion stuff will go away per our recent conversation.